File: v5/utils/network.py
"""
bilibili_api.utils.network

与网络请求相关的模块。能对会话进行管理（复用 TCP 连接）
"""

import aiohttp
from ..exceptions import ApiCodeException, ApiContentTypeException, NetworkException
import json
import re
from .Credential import Credential
from .. import settings

__session: aiohttp.ClientSession = aiohttp.ClientSession()


async def request(method: str, url: str, params: dict = None, data: dict = None,
                  credential: Credential = None, **kwargs):
    """
    向接口发送请求

    :param method: 请求方法
    :param url: 请求 URL
    :param params: 请求参数
    :param data: 请求载荷
    :param credential: Credential 类
    :return: 接口未返回数据时，返回 None，否则返回该接口提供的 data 或 result 字段的数据
    """
    method = method.upper()
    # 请求为非 GET 时要求 bili_jct
    if method != 'GET':
        credential.raise_for_no_bili_jct()

    # 使用 Referer 和 UA 请求头以绕过反爬虫机制
    DEFAULT_HEADERS = {
        "Referer": "https://www.bilibili.com/",
        "User-Agent": "Mozilla/5.0"
    }
    headers = DEFAULT_HEADERS

    if params is None:
        params = {}

    # Session 被关闭时，用于复用 TCP 连接
    global __session
    if __session.closed:
        __session = aiohttp.ClientSession()

    # 自动添加 csrf
    if method in ['POST', 'DELETE', 'PATCH']:
        if data is None:
            data = {}
        data['csrf'] = credential.bili_jct
        data['csrf_token'] = credential.bili_jct

    config = {
        "method": method,
        "url": url,
        "params": params,
        "data": data,
        "headers": headers,
        "cookies": credential.get_cookies()
    }

    config.update(kwargs)

    # 如果用户提供代理则设置代理
    if settings.proxy:
        config["proxy"] = settings.proxy

    resp = await __session.request(**config)

    # 检查状态码
    try:
        resp.raise_for_status()
    except aiohttp.ClientResponseError as e:
        raise NetworkException(e.status, e.message)

    # 检查响应头 Content-Length
    content_length = resp.headers.get("content-length")
    if content_length and int(content_length) == 0:
        return None

    # 检查响应头 Content-Type
    content_type = resp.headers.get("content-type")

    # 响应头不存在该字段
    if content_type is None:
        raise ApiContentTypeException("None")

    # 不是 application/json
    if content_type.lower().index("application/json") == -1:
        raise ApiContentTypeException(content_type)

    raw_data = await resp.text()
    resp_data: dict

    if 'jsonp' in params and 'callback' in params:
        # JSONP 请求
        resp_data = json.loads(re.match("^.*?({.*}).*$", raw_data, re.S).group(1))
    else:
        # JSON
        resp_data = json.loads(raw_data)

    # 检查 code
    code = resp_data.get("code", None)

    if code is None:
        raise ApiCodeException(-1, "API 返回数据未含 code 字段", resp_data)

    if code < 0:
        msg = resp_data.get('msg', None)
        if msg is None:
            msg = resp_data.get('message', None)
        if msg is None:
            msg = "接口未返回错误信息"
        raise ApiCodeException(code, msg, resp_data)

    real_data = resp_data.get("data", None)
    if real_data is None:
        real_data = resp_data.get("result", None)
    return real_data


async def close_session():
    """
    关闭请求 Session，在所有请求完毕后请务必调用这个方法
    :return: None
    """
    if not __session.closed:
        await __session.close()


async def upload_multipart(payload: dict):
    """
    上传 multipart/form-data 类型的数据

    :param payload: 要上传的数据
    """
    pass


def get_session():
    """
    获取当前模块的 aiohttp.ClientSession 对象，用于自定义请求
    """
    return __session

File: bilibili-api-ts/utils/network.ts
import { CookieJar } from 'tough-cookie';
import * as crypto from 'crypto';
import axios, { AxiosInstance } from 'axios';
import { wrapper } from 'axios-cookiejar-support';
import { Credential } from '../models/Credential';
import { Proxy } from '../models/Proxy';

const cookieJar = new CookieJar();

var sess: any = null;
var user_proxy: Proxy|null = null;

async function getAxiosInstance({credential=new Credential({}), proxy=null}: {credential?: Credential, proxy?: Proxy|null}) {
  if (credential.sessdata !== null) {
    cookieJar.setCookieSync(
      `SESSDATA=${credential.sessdata}; Domain=.bilibili.com`, 
      'https://www.bilibili.com'
    );
  }
  if (credential.bili_jct !== null) {
    cookieJar.setCookieSync(
      `bili_jct=${credential.bili_jct} Domain=.bilibili.com`, 
      'https://www.bilibili.com'
    );
  }
  if (credential.dedeuserid !== null) {
    cookieJar.setCookieSync(
      `DedeUserID=${credential.dedeuserid}; Domain=.bilibili.com`, 
      'https://www.bilibili.com'
    );
  }
  cookieJar.setCookieSync(
     `buvid3=${crypto.randomUUID()}; Domain=.bilibili.com`,
     'https://www.bilibili.com'
  );
  // console.log(cookieJar.getCookieString("https://api.bilibili.com"));
  sess = wrapper(
    axios.create({
      headers: {
        'user-agent': 'Mozilla/5.0',
        referer: 'https://www.bilibili.com/',
      },
      // withCredentials: true, 
      jar: cookieJar,
      responseType: 'json',
      proxy: proxy
        ? {
          host: proxy.hostname,
          port: parseInt(proxy.port),
          auth:
              proxy.username && proxy.password
                ? {
                  username: proxy.username,
                  password: proxy.password,
                }
                : undefined,
        }
        : false,
    }));

    return sess;
};

async function setAxiosInstance(axios: AxiosInstance) {
  sess = axios;
}

export async function get_session({credential=new Credential({})}: {credential: Credential}) {
  if (credential === null || credential === undefined) {
    credential = new Credential({});
  }
  if (user_proxy !== null) {
    return await getAxiosInstance({credential: credential, proxy: user_proxy});
  }
  else {
    return await getAxiosInstance({credential: credential});
  }
}


export async function request(
  {
    method, url, params={}, data={}, credential=new Credential({}), no_csrf=false
  } : 
  {
    method: string, 
    url: string,
    params?: any, 
    data?: any, 
    credential?: Credential|null, 
    no_csrf?: boolean
  }
) {
  if (no_csrf === null || no_csrf === undefined) {
    no_csrf = false;
  }

  if (credential === null || credential === undefined) {
    credential = new Credential({});
  }

  method = method.toUpperCase();

  if (method !== "GET" && !no_csrf){
    credential.raise_for_no_bili_jct({})
  }

  const DEFAULT_HEADERS = {
    "Referer": "https://www.bilibili.com",
    "User-Agent": "Mozilla/5.0",
  };
  var headers = DEFAULT_HEADERS;

  if (!no_csrf && method in ["POST", "DELETE", "PATCH"]){
    if (data === null){
      data = {};
    }
    data["csrf"] = credential.bili_jct;
    data["csrf_token"] = credential.bili_jct;
  }

  if (params === null) {
    params = {};
  }

  if (params['jsonp'] !== undefined) {
    params["callback"] = "callback"
  }

  var cookies = credential.get_cookies();

  var config: Record<string, any> = {
    "method": method, 
    "url": url, 
    "params": params, 
    "data": data, 
    "headers": headers, 
    "cookies": cookies
  }

  // console.log(credential.get_cookies());

  if (user_proxy !== null) {
    await getAxiosInstance({credential: credential, proxy: user_proxy});
  }
  else {
    await getAxiosInstance({credential: credential});
  }
  var resp = await (await sess).request(config);

  // console.log(resp)

  // console.log(resp.headers)

  var has_content_type = resp.headers['content-type'] ? true : false;
  if (!has_content_type) return;
  if (resp.headers['content-type'].toLowerCase().indexOf("application/json") === -1){
    throw "响应不是 application/json 类型";
  }

  var resp_data = resp['data'];
  // console.log(resp_data);
  // console.log(JSON.stringify(raw_data))
  var code = resp_data['code'];
  if (code === null || code === undefined) {
    throw "API 返回数据未含 code 字段";
  }
  if (code !== 0) {
    var msg = resp_data['message'];
    if (msg === undefined) {
      msg = "接口未返回错误信息";
    }
    throw code + msg;
  }

  var real_data = resp_data['data'];
  if (real_data === undefined) {
    real_data = resp_data['result'];
  }
  return real_data;
}

export function set_proxy(config: any){
  user_proxy = config.proxy;
}
